src/window.cpp

```diff
@@ -3060,22 +3059,61 @@ void InputLoop()
 
 	/* HandleMouseEvents was already called for this tick */
 	HandleMouseEvents();
-	HandleAutoscroll();
+}
+
+/**
+ * Dispatch OnRealtimeTick event over all windows
+ */
+void CallWindowRealtimeTickEvent(uint delta_ms)
+{
+	Window *w;
+	FOR_ALL_WINDOWS_FROM_FRONT(w) {
+		w->OnRealtimeTick(delta_ms);
+	}
 }
 
 /**
  * Update the continuously changing contents of the windows, such as the viewports
  */
 void UpdateWindows()
 {
+	static uint32 last_realtime_tick = _realtime_tick;
+	uint delta_ms = _realtime_tick - last_realtime_tick;
+	last_realtime_tick = _realtime_tick;
+
+	if (delta_ms == 0) return;
+
+	CallWindowRealtimeTickEvent(delta_ms);
+
+	static int network_message_timer = 1;
+	if (TimerElapsed(network_message_timer, delta_ms)) {
+		network_message_timer = 1000;
+		NetworkChatMessageLoop();
+	}
+
 	Window *w;
 
+	static int window_timer = 1;
+	if (TimerElapsed(window_timer, delta_ms)) {
+		window_timer = MILLISECONDS_PER_TICK;
+
+		extern int _caret_timer;
+		_caret_timer += 3;
+		CursorTick();
+
+		HandleKeyScrolling();
+		HandleAutoscroll();
+		DecreaseWindowCounters();
+	}
+
 	static int highlight_timer = 1;
-	if (--highlight_timer == 0) {
-		highlight_timer = 15;
+	if (TimerElapsed(highlight_timer, delta_ms)) {
+		highlight_timer = 450;
```

Comment:
magic number make me sad

Reply:
450 ms. Should it be a constant, or just a comment?